codeworld-error-sanitizer/src/RegexShim.hs

```diff
@@ -38,23 +39,45 @@ replace regex replacement str = textFromJSString $
 
 #else
 
-import Data.Array (elems)
+import Data.Array (bounds, elems, (!))
+import Data.Char (isDigit)
 import Data.List
 import Data.Monoid
 import qualified Data.Text as T
+import Data.Text.Read (decimal)
 import Text.Regex.Base
 import Text.Regex.TDFA
 import Text.Regex.TDFA.Text
 
+getInserts :: Text -> [(Int, (Int, Int))]
+getInserts replacement = case replacement =~ ("[\\][0-9]" :: Text) :: [MatchArray] of
+    [] -> []
+    groups -> fmap processInsertion groups
+    where processInsertion group =
+              case decimal $ T.take len $ T.drop (start + 1) replacement of
+                  Right (index, _) -> (index, (start, len))
+                  Left _ -> error $ T.unpack $ "Can not parse replacement groups in "
+                                  <> replacement -- should never happen beause handled
+                                                 -- by regex matching above
+              where (start, len) = group ! 0
+
+applyInserts :: Text -> Text -> MatchArray -> [(Int, (Int, Int))] -> Text
```

Comment:
I don't think this works for more than one group, because you need to fix up the remaining indices in the recursive case.  You might try something like this instead.

    type Inserts = [(Int, (Int, Int))]
    applyInserts :: Text -> Text -> MatchArray -> Inserts -> Text
    applyInserts str replacement matches inserts = go 0 replacement inserts
      where go pos replacement ((index, (rstart, rlen)) : inserts) =
                let pre = T.take (rstart - pos) replacement
                    rest = T.drop (rlen + rstart - pos) replacement
                    (sstart, slen) = matches ! index
                    source = T.take slen $ T.drop sstart str
                in pre <> source <> go (pos + rstart + rlen) rest inserts
            go _ replacement [] = replacement